packages/gatsby/src/internal-plugins/functions/gatsby-node.ts

```diff
@@ -470,6 +470,13 @@ export async function onCreateDevServer({
 
       return next()
     },
+    (req, res, next) => {
+      const ALLOW_REGEX = /^x-gatsby-(A-Za-z)*/
+
+      req.headers = _.pickBy(req.headers, (value, key) => ALLOW_REGEX.test(key))
+
+      return next()
+    },
```

Comment:
we can use simple `startsWith` check which is cheaper than regex
```suggestion
    (req, res, next) => {
      req.headers = _.pickBy(req.headers, (value, key) => key.toLowerCase().startsWith('x-gatsby-'))

      return next()
    },
```